share/pvss/pvss.go

```diff
@@ -152,39 +166,94 @@ func DecShareBatch(suite Suite, H kyber.Point, X []kyber.Point, sH []kyber.Point
 // log_{G}(X) == log_{sG}(sX). Note that X = xG and sX = s(xG) = x(sG).
 func VerifyDecShare(suite Suite, G kyber.Point, X kyber.Point, encShare *PubVerShare, decShare *PubVerShare) error {
 	if err := decShare.P.Verify(suite, G, decShare.S.V, X, encShare.S.V); err != nil {
-		return errorDecVerification
+		return ErrDecVerification
 	}
 	return nil
 }
 
 // VerifyDecShareBatch provides the same functionality as VerifyDecShare but for
 // slices of decrypted shares. The function returns the the valid decrypted shares.
-func VerifyDecShareBatch(suite Suite, G kyber.Point, X []kyber.Point, encShares []*PubVerShare, decShares []*PubVerShare) ([]*PubVerShare, error) {
-	if len(X) != len(encShares) || len(encShares) != len(decShares) {
-		return nil, errorDifferentLengths
+func VerifyDecShareBatch(suite Suite, G kyber.Point, pubRun *PubVSSData, decShares []*PubVerShare, cExp kyber.Scalar) ([]*PubVerShare, error) {
+	// match index to encShare and decShare
+	indToShare := make(map[int]*PubVerShare, pubRun.N)
+	indToPk := make(map[int]kyber.Point, pubRun.N)
+	for i, el := range pubRun.Shares {
+		indToShare[el.S.I] = el
+		indToPk[el.S.I] = pubRun.X[i]
+	}
+
+	filteredDecShares := make([]*PubVerShare, 0)
+	filteredEncShares := make([]*PubVerShare, 0)
+	filteredPks := make([]kyber.Point, 0)
+	for _, decShare := range decShares {
+		if encShare, ok := indToShare[decShare.S.I]; ok {
+			filteredPks = append(filteredPks, indToPk[decShare.S.I])
+			filteredEncShares = append(filteredEncShares, encShare)
+			filteredDecShares = append(filteredDecShares, decShare)
+		}
 	}
+
+	if len(filteredPks) != len(filteredEncShares) || len(filteredEncShares) != len(filteredDecShares) {
+		return nil, ErrDifferentLengths
+	}
+
 	var D []*PubVerShare // good decrypted shares
-	for i := 0; i < len(X); i++ {
-		if err := VerifyDecShare(suite, G, X[i], encShares[i], decShares[i]); err == nil {
-			D = append(D, decShares[i])
+	for i := 0; i < len(filteredPks); i++ {
+		if err := VerifyDecShare(suite, G, filteredPks[i], filteredEncShares[i], filteredDecShares[i]); err != nil {
+			continue
+		}
+		if !pubRun.Shares[i].P.C.Equal(cExp) {
```

Comment:
I find this very surprising that all shares are sharing the same challenge, and also shouldn't you rather be checking the response ?

Each of them should have a different challenge, and you should instead be using `dleq.Verify` to verify the proof, which requires knowing `G`, `H`,  `xG` and `xH`, as well as the proof.

See here:
https://asecuritysite.com/encryption/go_dleq

Please can you add a test to `dleq_test.go` using the values on this page to check it does work as expected?